### PR TITLE
Fix string make_key

### PR DESCRIPTION
**Pull request by jeffreyssmith2nd**
## Change Description

Fixes the string make_key implementation.

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 1]
```diff
@@ -254,17 +254,35 @@ inline key_type make_key(double val) {
...
 
-inline key_type make_key(const char* str, size_t size, bool case_insensitive=false) {
-   size_t data_size = size + 3;
+inline key_type make_key(const char* val, size_t size, bool case_insensitive=false) {
+   size_t num_zeroes = 0;
+   for (int i = 0; i < size; ++i) {
+      if (val[i] == 0x00) {
+         ++num_zeroes;
+      }
+   }
```
Comment (swatanabe-b1) [inline libraries/eosiolib/contracts/eosio/key_value.hpp]: std::count(val, val+size, 0)

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 2]
```diff
@@ -254,17 +254,35 @@ inline key_type make_key(double val) {
...
+
+   size_t data_size = size + num_zeroes + 2;
    void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
```
Comment (swatanabe-b1) [inline libraries/eosiolib/contracts/eosio/key_value.hpp]: Cast this to ``char*`` here, so you don't need all the casts later.

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 3]
```diff
@@ -254,17 +254,35 @@ inline key_type make_key(double val) {
...
 
-   if (case_insensitive) {
-      std::transform(str, str + size, (char*)data_buffer, [](unsigned char c) -> unsigned char { return std::toupper(c); });
-   } else {
-      memcpy(data_buffer, str, size);
+   int j = 0;
+   for(int i = 0; i < size; ++i) {
```
Comment (swatanabe-b1) [inline libraries/eosiolib/contracts/eosio/key_value.hpp]: Make sure that your types match.  size is a `size_t`

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 4]
```diff
@@ -254,17 +254,35 @@ inline key_type make_key(double val) {
...
+      if (val[i] == 0x00) {
+         ((char*)data_buffer)[j] = 0x00;
+         ++j;
```
Comment (swatanabe-b1) [inline libraries/eosiolib/contracts/eosio/key_value.hpp]: Post-increment is less verbose: ``data_buffer[j++]``

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 5]
```diff
@@ -254,17 +254,35 @@ inline key_type make_key(double val) {
    return make_floating_key<uint64_t>(val);
 }
 
-inline key_type make_key(const char* str, size_t size, bool case_insensitive=false) {
-   size_t data_size = size + 3;
+inline key_type make_key(const char* val, size_t size, bool case_insensitive=false) {
```
Comment (swatanabe-b1) [inline libraries/eosiolib/contracts/eosio/key_value.hpp]: I thought we were going to kill case insensitive comparison?

Comment (jeffreyssmith2nd) [reply libraries/eosiolib/contracts/eosio/key_value.hpp]: I was under the impression we were going to kill it for unicode but still support it for ASCII, but maybe that's not accurate.

Comment (swatanabe-b1) [reply libraries/eosiolib/contracts/eosio/key_value.hpp]: I think the final resolution was to remove it entirely.

**Comment by jeffreyssmith2nd** [comment]
Handled by #850

**Pull request closed without merging**